Many zabbix inventory handling fixes.

piterpunk opened the pull request:
### What does this PR do?
- Adds `inventory_clean` parameter to select if the current inventory should or not be overwritten;
- Allows inventory to be passed as a list or dictionary;
- Don't remove the current inventory if `inventory` parameter is missing;
- Compare the key values and not only the keys itself to update inventory;
- Update documentation with the new parameter and dictionary change;
- Restore `inventory_mode` to the original value after inventory update;
- Allows to erase an inventory field content by sending it as an empty string.
- Fixes an exception when `zabbix.host_inventory_get` is called for a non-existent host.

### What issues does this PR fix or reference?
Fixes: #60389
Fixes: #60382
Fixes: #60418 

### Previous Behavior
Current inventory was always erased and overwritten with the declared in `inventory` parameter. Missing `inventory` parameter erases all the current inventory and after inventory changes the `inventory_mode`is always changed to `'0'`

When `zabbix.host_inventory_get` is called for a missing host it throws an exception instead returns `False`

### New Behavior
Default behaviour is still to overwrite the current inventory with the data in `inventory` parameter, but this behaviour can be controlled by `inventory_clean` parameter (which defaults to True). A missing `inventory` parameter now makes the inventory to be ignored, if the user wants an empty inventory it can sends `inventory_mode: '-1'` or `inventory: {}`.

Now `zabbix.host_inventory_get` returns `False` when called for a non-existent host.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Commit:
Various zabbix.host_present inventory fixes

- Adds inventory_clean parameter to select if the current inventory
  should or not be overwrited
- Allows inventory to be passed as a list or dictionary
- Don't remove the current inventory if inventory parameter is missing
- Compare the key values and not only the keys itself to update inventory
- Update documentation with the new parameter and dictionary change

OrangeDog commented:
> `inventory_mode: -1`

It has to be a string, so `inventory_mode: '-1'`. Or the module could `str()` every parameter before building the API request.

Commit:
Restore the inventory_mode after inventory update

Commit:
Adds zabbix_host.present tests and changelog

Commit:
Adjusts zabbix_host.present new tests to Python3.5

OrangeDog commented on salt/states/zabbix_host.py:
```diff
@@ -397,12 +415,19 @@ def _interface_format(interfaces_data):
                 ret["changes"]["inventory"] = str(new_inventory)
                 if "error" in hostupdate:
                     error.append(hostupdate["error"])
+                # restore inventory_mode as host_inventory_set changes it to '0'
```
Can `host_inventory_set` not be changed to take an `inventory_mode`?

piterpunk replied:
Yes, it could, but I guess it's a state related statement and the state module is a better place for it.

If you are thinking about avoid send this additional `host_update`, it can´t be done as the `inventory_mode` needs to be set to `0` to being able to set the inventory. If the `inventory_mode` is `1` and you try to manually set the inventory you'll get a permission error:

`Zabbix API: Invalid params. (No permissions to referred object or it does not exist!)`

So, as there is a need of that additional `host_update` in someplace and its use is to preserve the `inventory_mode` state, I thought the code should be in the state module instead in the execution module.

OrangeDog replied:
It might depend on the version of Zabbix. Older releases had a lot of bugs with inventory in the API.

piterpunk replied:
> It might depend on the version of Zabbix. Older releases had a lot of bugs with inventory in the API.

Maybe, but my test error was due lack of attention. I removed and recreated the host but send the tests with the old ID. Of course the `referred object ... does not exist!`.

piterpunk replied:
Done

OrangeDog commented on salt/states/zabbix_host.py:
```diff
@@ -397,12 +415,19 @@ def _interface_format(interfaces_data):
                 # combine connection_args, inventory, and clear_old
                 sum_kwargs = dict(new_inventory)
```
Is already a `dict`.

piterpunk replied:
Good catch. I'll change it.

piterpunk replied:
Done

piterpunk replied:
Ooops, the `dict` in `dict(new_inventory)` is there to copy `new_inventory` to `sum_kwargs`. Changed to `new_inventory.copy()` so this line will not induce other people to this same error and remove it again.

Commit:
Adds inventory_mode param to host_inventory_set

- Allows inventory_mode parameter to zabbix.host_inventory_set
- Changes zabbix_host.present to call host_inventory_set with the
  aforementioned parameter
- Updates the tests for zabbix and zabbix_host modules.

Commit:
Fixes zabbix_host.present errors caught on tests

piterpunk commented:
re-run pr-windows-2019-x64-py3-pytest

OrangeDog commented:
I don't see a test case for removing inventory items. e.g.
```
- inventory_clean: false
- inventory:
    asset_tag: abc
    vendor: ''
```
should change `{"alias": "foo", "vendor":"bar"}` into `{"alias": "foo", "asset_tag": "abc"}`

piterpunk commented:
> I don't see a test case for removing inventory items. e.g.
> ...
> should change `{"alias": "foo", "vendor":"bar"}` into `{"alias": "foo", "asset_tag": "abc"}`

Did a quick (manual) test here and already catch a problem. While it sends the correct call to Zabbix there is something wrong in comparison and the state is always 'updated'
```
local:
----------
          ID: jamanta
    Function: zabbix_host.present
      Result: True
     Comment: Host jamanta updated.
     Started: 11:46:52.340137
    Duration: 2470.918 ms
     Changes:   
              ----------
              inventory:
                  {'asset_tag': '', 'vendor': 'Teste'}

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   2.471 s
```
I'll fix and write a test case for it.

Commit:
Clear an inventory key sending it as empty string

- Allow sending an empty string as key value to erase it from inventory
- Add test to this use case

Commit:
Fix exception on zabbix.host_inventory_get

piterpunk commented:
Ok, I guess it's now all done.

Included a fix for #60418 as it was a simple change to do, related to Zabbix inventory and we're already touching `salt/modules/zabbix.py`

piterpunk commented:
re-run  pr-centos-7-x86_64-py3-m2crypto-pytest

piterpunk commented:
@sagetherage my last big batch of changes for zabbix (#58604) did wait a lot of time before get merged. I know that you have a lot of work to do, so I am doing this heads up to prevent the same happens again. This PR closes three issues and would be nice to have it evaluated and merged.

Thank you!

piterpunk commented:
re-run pr-debian-11-amd64-py3-pytest

waynew commented on salt/modules/zabbix.py:
```diff
@@ -1430,7 +1430,7 @@ def host_inventory_get(hostids, **connection_args):
             return (
                 ret["result"][0]["inventory"]
-                if len(ret["result"][0]["inventory"]) > 0
+                if ret["result"] and ret["result"][0]["inventory"]
```
**comment**: I don't think it reads any better, but it *could* be written:

```
return next(iter(ret.get("result") or [{"inventory": False}])).get("inventory") or False
```

(or, something like that. But, like I said, it's not exactly an improvement :trollface: )

piterpunk replied:
Oh, of course I have a bias here, but this one I think my version is easier to understand.

twangboy replied:
I like "easier to understand"

waynew commented on salt/modules/zabbix.py:
```diff
@@ -1471,8 +1471,10 @@ def host_inventory_set(hostid, **connection_args):
-
             connection_args.pop("clear_old", None)
+
+            inventory_mode = connection_args.pop("inventory_mode", "0")
```
**question**: does inventory_mode need to be removed from connection_args?

piterpunk replied:
The `connection_args`, despite its name, carriers all kind of information, not only connection arguments like server, user and password. The inventory items are inside `connection_args` and needs to be moved to the element `inventory` before sent to Zabbix API, the `inventory_mode` is also inside `connection_args` but it's not an inventory item, it needs to be at same level of `inventory`, not inside it.

So, the parameters needs to be moved around to transform a flat set of parameters in the hierarchy expected by `host.update`. The `inventory_mode` pop is part of that parameters dance.

waynew commented on salt/states/zabbix_host.py:
```diff
@@ -33,12 +34,15 @@ def present(host, groups, interfaces, **kwargs):
     :param visible_name: Optional - string with visible name of the host, use 'visible_name' instead of 'name' \
     parameter to not mess with value supplied from Salt sls file.
+    :param inventory_clean: Optional - Boolean value that selects if the current inventory will be cleaned and \
+    overwritten by the declared inventory list (True); or if the inventory will be kept and only updated with \
```
**question** are the backslashes necessary?

waynew replied:
(I see they're elsewhere... I'm just pretty sure that sphinx will remove them :shrug: )

piterpunk replied:
I don't know if they are needed, I only followed the `:param visible_name` that was already there.

piterpunk replied:
Removed the backslashes.

waynew commented on salt/states/zabbix_host.py:
```diff
@@ -291,6 +299,13 @@ def _interface_format(interfaces_data):
         if host_updated_params:
             update_host = True
 
+        if "inventory_mode" in host_extra_properties:
+            inventory_mode = host_extra_properties["inventory_mode"]
+        elif host["inventory_mode"] == "-1":
+            inventory_mode = "0"
+        else:
+            inventory_mode = host["inventory_mode"]
+
```
**question** would this be clearer written like:

```
inventory_mode = host_extra_properties.get("inventory_mode", host["inventory_mode"])
inventory_mode = "0" if inventory_mode == "-1" else inventory_mode
```
?

waynew replied:
(looking at the below code - it might be better as:

```
host_mode = host["inventory_mode"]
inventory_mode = host_extra_properties.get("inventory_mode", "0" if host_mode == "-1" else host_mode)
```

piterpunk replied:
Looks good and better to understand. Any problem if I eliminate the `host_mode` and just use `host["inventory_mode"]` in the `inventory_mode` line, or change it to `host_inventory_mode`? `host_mode` looks easy to be mistaken as a variable related to `host` status and not limited to inventory scope.

piterpunk replied:
Made the change, but instead `host_mode` I did use `host_inventory_mode`

waynew commented on salt/states/zabbix_host.py:
```diff
@@ -336,21 +351,33 @@ def _interface_format(interfaces_data):
-            update_inventory = True
+        # if inventory param is empty leave inventory as is don't compare it
+        # if inventory_mode is '-1', the inventory will be erased, why compare it?
+        if inventory is not None and inventory_mode != "-1":
```
**question** aren't we making `inventory_mode` something else? Or is that only when the *host* provides -1?

piterpunk replied:
Yes... I guess this is becoming more confused as time is passing.

This is the `inventory_mode` from state declaration. The override of `-1` is only if the `host` provides it. This override is just to pass this test and don't compare the inventory if it's set to be erased (`-1` in state declaration) but compare inventory if it's currently disabled (`-1` in host) and being populated now.

piterpunk replied:
Did the change in `inventory_mode` handling. Hope it's now more comprehensible.

waynew replied:
:+1: This seems to be reasonable

waynew commented on salt/states/zabbix_host.py:
```diff
@@ -336,21 +351,33 @@ def _interface_format(interfaces_data):
+                hostids=hostid, **connection_args
+            )
+            if cur_inventory:
+                cur_inventory.pop("hostid", None)
+                cur_inventory.pop("inventory_mode", None)
+                if new_inventory:
+                    for key in new_inventory:
+                        if (
+                            key in cur_inventory
+                            and new_inventory[key] != cur_inventory[key]
+                        ):
+                            update_inventory = True
+                            break
+                else:
+                    # remove the empty keys in current inventory
+                    cur_inventory = {k: v for k, v in cur_inventory.items() if v}
+                    # if current inventory any no empty key and new inventory is empty
+                    if not cur_inventory:
+                        update_inventory = True
```
**suggestion**: check out [`salt.utils.dictdiffer`](https://github.com/saltstack/salt/blob/master/salt/utils/dictdiffer.py), it's built for this purpose

piterpunk replied:
I'll check it.

piterpunk replied:
Nice, the code was changed to use `salt.utils.dictdiffer.diff()`

waynew reviewed (commented):
Just a few comments - I haven't finished my review/checked out the tests yet, but it's looking good so far :+1: I'll see about finishing my review either later or tomorrow

Commit:
Zabbix inventory handling enhancements and fixes

- Remove the backslashes in documentation
- Streamline the inventory_mode attribution
- Use salt.utils.dictdiffer.diff() to compare the inventories
  dictionaries

piterpunk commented:
In addition of the asked changes, I did change the `zabbix.host_inventory_get` first parameter from `hostids` to `hostid` and update its description. As the method  was already returning the inventory of only one host, the `hostids` name, seems to be a little misleading, while being the parameter name of  Zabbix API call.

piterpunk commented:
re-run pr-ubuntu-1804-amd64-py3-pytest

OrangeDog commented:
> change the `zabbix.host_inventory_get`  first parameter from `hostids` to `hostid`

That's a breaking change. It would have to support both and go through a deprecation cycle.

Commit:
Revert parameter name change in host_inventory_get

piterpunk commented:
> > change the `zabbix.host_inventory_get`  first parameter from `hostids` to `hostid`
> That's a breaking change. It would have to support both and go through a deprecation cycle.

Oh, Ok. I just reverted the parameter name change, but kept the description change. With this, we avoid the breaking change but the documentation will be clear about the working.

piterpunk commented:
re-run pr-windows-2019-x64-py3-pytest

piterpunk commented:
re-run pr-centos-7-x86_64-py3-pycryptodome-pytest

sagetherage commented:
@piterpunk 
we are working the Centos7 test failures today :)

piterpunk commented:
@waynew I guess that all review points are solved now, can you confirm it?

piterpunk commented:
re-run pr-ubuntu-2004-amd64-py3-tcp-pytest

waynew commented on salt/modules/zabbix.py:
```diff
@@ -1398,14 +1398,14 @@ def host_update(hostid, **connection_args):
 
 
-def host_inventory_get(hostids, **connection_args):
+def host_inventory_get(hostid, **connection_args):
```
**question** Does this affect existing zabbix users? If so we'll want to deprecate the hostids (it would still be in connection_args)

waynew replied:
whoops, see that was reverted :joy:

waynew reviewed (changes requested):
This is looking good! I have a couple of points around tests and some questions - feel free to ping me next week if I haven't re-reviewed this :+1:

OrangeDog commented on salt/states/zabbix_host.py:
```diff
@@ -395,9 +405,10 @@ def _interface_format(interfaces_data):
             if update_inventory:
                 # combine connection_args, inventory, and clear_old
-                sum_kwargs = dict(new_inventory)
+                sum_kwargs = new_inventory.copy()
```
Inventory shouldn't have any nesting - it's one flat dict.
A deep copy wouldn't hurt though.

piterpunk replied:
Yes yes, as invenctory is a flat dict I preferred to use `copy` instead `deepcopy`. But probably change to `deepcopy` is a good idea.

piterpunk commented on tests/unit/modules/test_zabbix.py:
```diff
@@ -1124,12 +1201,65 @@ def test_host_inventory_set(self):
         query_return = {"jsonrpc": "2.0", "result": {"hostids": [10258]}, "id": 0}
-
-        with patch.object(zabbix, "_query", return_value=query_return):
+        mock__query = MagicMock(return_value=query_return)
```
Hmmm my idea for this name was:
```
mock + separator + function =
mock + _              + _query =
mock__query
```
That is why it have double underscore.

piterpunk commented on tests/unit/modules/test_zabbix.py:
```diff
@@ -1124,12 +1201,65 @@ def test_host_inventory_set(self):
-
-        with patch.object(zabbix, "_query", return_value=query_return):
+        mock__query = MagicMock(return_value=query_return)
+        with patch.object(zabbix, "_query", mock__query):
```
I'll change it and test tonight

Commit:
Minor adjusts in zabbix unit tests

waynew reviewed (commented):
re-running the failed Windows test... I *think* it might be a flaky test failure... we'll see. Otherwise, I think this looks good.

piterpunk committed:
Merge branch 'master' into zabbix.host.present_keep_inventory-Fix_Issue_60389

piterpunk commented:
re-run pr-macosx-mojave-x86_64-py3-pytest

piterpunk commented:
re-run pr-freebsd-130-amd64-py3-pytest

waynew commented:
re-run macosx

piterpunk committed:
Merge branch 'master' into zabbix.host.present_keep_inventory-Fix_Issue_60389

piterpunk commented:
@sagetherage can this one be merged for Silicon?

garethgreenaway committed:
Merge branch 'master' into zabbix.host.present_keep_inventory-Fix_Issue_60389

Ch3LL committed:
Merge branch 'master' into zabbix.host.present_keep_inventory-Fix_Issue_60389
